fl4health/clients/moon_client.py

```diff
@@ -0,0 +1,107 @@
+import copy
+from pathlib import Path
+from typing import Dict, List, Optional, Sequence
+
+import torch
+from flwr.common.typing import Config, NDArrays
+
+from fl4health.checkpointing.checkpointer import TorchCheckpointer
+from fl4health.clients.basic_client import BasicClient
+from fl4health.model_bases.moon_base import MoonModel
+from fl4health.utils.losses import Losses, LossMeterType
+from fl4health.utils.metrics import Metric, MetricMeterType
+
+
+class MoonClient(BasicClient):
+    """
+    This client implements the MOON algorithm from Model-Contrastive Federated Learning. The key idea of MOON
+    is to utilize the similarity between model representations to correct the local training of individual parties,
+    i.e., conducting contrastive learning in model-level.
+    """
+
+    def __init__(
+        self,
+        data_path: Path,
+        metrics: Sequence[Metric],
+        device: torch.device,
+        loss_meter_type: LossMeterType = LossMeterType.AVERAGE,
+        metric_meter_type: MetricMeterType = MetricMeterType.AVERAGE,
+        checkpointer: Optional[TorchCheckpointer] = None,
+    ) -> None:
+        super().__init__(
+            data_path=data_path,
+            metrics=metrics,
+            device=device,
+            loss_meter_type=loss_meter_type,
+            metric_meter_type=metric_meter_type,
+            checkpointer=checkpointer,
+        )
+        self.initial_tensors: List[torch.Tensor]
+        self.contrastive_weight: float = 10
+        self.temprature: float = 0.5
+        self.current_loss: float
+        self.cos_sim = torch.nn.CosineSimilarity(dim=-1)
+        self.ce_criterion = torch.nn.CrossEntropyLoss().to(self.device)
+        self.len_old_models_buffer: int = 1
+        self.old_models_list: list[MoonModel] = []
+        self.global_model: MoonModel
+        self.features: torch.Tensor
+        self.old_features_list: list[torch.Tensor] = []
+        self.global_features: torch.Tensor
+
+    def predict(self, input: torch.Tensor) -> Dict[str, torch.Tensor]:
+        preds, self.features, _ = self.model(input)
+        self.features = self.features.view(len(self.features), -1)
+        self.old_features_list = []
+        for old_model in self.old_models_list:
```

Comment:
In order to avoid having this side effect (setting global_features and old_features_list attribute in predict method), we can stack the list of old_features into a tensor and include it in the preds dict. Would also include global_features in the preds dict. Then get_contrastive loss would accept preds. @emersodb do you think is a worthwhile refactor, or is it better this way?

Reply:
I agree this can be cleaner. I will apply it in refactor now

Reply:
Oh but no, I get into errors with metrics file as it expects only predictions.

Reply:
Good catch! I think the way you did it is perfect. We should be able to have predictions that we don't have to run metrics on. I am going to update the MetricMeterManager in a subsequent PR so its easier for users to specify which predictions they want evaluated. This is a good first step, thanks Sana :)